server/src/main/java/mineclone/server/world/ServerWorld.java

```diff
@@ -125,164 +133,206 @@ public void growTree(IBlockPosition pos) {
 			if (i < treeHeight - 1) {
 				tmpPos.x = pos.getX();
 				tmpPos.z = pos.getZ();
-				
-				setBlock(tmpPos, Blocks.LOG_BLOCK, true);
+
+				setBlock(tmpPos, Blocks.LOG_BLOCK, SetBlockFlags.UPDATE_NEIGHBOR_SHAPES | SetBlockFlags.UPDATE_NEIGHBORS);
 			}
 
 			tmpPos.y++;
 		}
 	}
-	
+
 	@Override
 	public boolean setBlockState(IBlockPosition pos, IBlockState newState) {
-		return setBlockState(pos, newState, true);
+		return setBlockState(pos, newState, SetBlockFlags.UPDATE_NEIGHBOR_SHAPES | SetBlockFlags.UPDATE_NEIGHBORS);
 	}
-	
+
 	@Override
-	public boolean setBlockState(IBlockPosition pos, IBlockState newState, boolean updateNeighbors) {
+	public boolean setBlockState(IBlockPosition pos, IBlockState newState, int flags) {
 		IBlockState oldState = chunkManager.getBlockState(pos);
 
 		if (chunkManager.setBlockState(pos, newState)) {
-			if (updateNeighbors) {
-				Block oldBlock = oldState.getBlock();
-				Block newBlock = newState.getBlock();
-				
-				if (oldBlock == newBlock) {
-					newBlock.onStateChanged(this, pos, oldState, newState);
-				} else {
-					oldBlock.onBlockRemoved(this, pos, oldState);
-					newBlock.onBlockAdded(this, pos, newState);
-				}
+			IBlock oldBlock = oldState.getBlock();
+			IBlock newBlock = newState.getBlock();
+
+			if (oldBlock == newBlock) {
+				newBlock.onChanged(this, pos, oldState, newState);
+			} else {
+				oldBlock.onRemoved(this, pos, oldState);
+				newBlock.onAdded(this, pos, newState);
+
+				tryUpdateNeighborShapes(pos, oldState, flags);
```

Comment:
Why are updated duplicated? Surely it can be solved in the update methods of the blocks, such that it properly handles changes when a block is replaced by another.

I am open for a discussion on this.

Reply:
tbh I wasn't completely happy with it either. I wanted to be able to control the updates with set block flags, but on the other hand handling them in the blocks' `onAdded` and `onRemoved` methods seems cleaner.